src/server.rs

```diff
@@ -1,23 +1,159 @@
 
 pub struct Server {
-    remote: remote::Remote,
+    remote_receiver: mpsc::Receiver<(String, Bytes)>,
+    topic: String,
+    local_addr: Ipv4Addr,
+    id_length: usize,
+    ip_pool: Arc<Mutex<IpPool>>,
+    sink: TunSink,
 }
 
 impl Server {
     pub fn new(config: config::Config) -> Self {
         let mqtt_options = config.broker_mqtt_options();
         let server_config = config.server.expect("Server config to be non-null");
 
-        let remote = remote::Remote::new(&mqtt_options, vec![server_config.topic]);
-        Self { remote: remote }
+        let (remote, remote_receiver) =
+            remote::Remote::new(&mqtt_options, vec![server_config.topic.clone()]);
+        let netmask = config.driver.tun.netmask;
+        let ip_network = Ipv4Network::with_netmask(server_config.bind_addr, netmask)
+            .expect("A proper ip network");
+        let bind_addr = ip_network.ip();
+
+        log::info!("Binding to {:?} netmask {:?}", bind_addr, netmask);
+
+        let ip_iter = SizedIpv4NetworkIterator::new(ip_network);
+
+        let mut tun_config = tun::Configuration::default();
+        tun_config.address(bind_addr);
+        tun_config.destination(bind_addr);
+        tun_config.netmask(netmask);
+
+        #[cfg(target_os = "linux")]
+        tun_config.platform(|tun_config| {
+            tun_config.packet_information(true);
+        });
+
+        tun_config.up();
+
+        let dev = tun::create_as_async(&tun_config).expect("Tunnel");
+        let (sink, mut stream) = dev.into_framed().split();
+
+        let ip_pool_arc = Arc::new(Mutex::new(LookupPool::new(ip_iter)));
+        let loop_ip_pool_arc = ip_pool_arc.clone();
+
+        task::spawn(async move {
+            while let Some(packet) = stream.next().await {
+                match packet {
+                    Ok(pkt) => {
+                        let result =
+                            Self::handle_packet(&remote, loop_ip_pool_arc.clone(), &pkt).await;
+                        if let Err(err) = result {
+                            log::error!("Error in handle_packet {:?}", err);
+                        }
+                    }
+                    Err(err) => panic!("Cannot get packet from tun, error: {:?}", err),
+                }
+            }
+        });
+
+        Self {
+            remote_receiver,
+            topic: server_config.topic,
+            local_addr: config.driver.local_addr,
+            id_length: server_config.id_length,
+            ip_pool: ip_pool_arc,
+            sink,
+        }
+    }
+
+    // tun -> mqtt
+    async fn handle_packet(
+        remote: &remote::Remote,
+        ip_pool: Arc<Mutex<IpPool>>,
+        packet: &TunPacket,
+    ) -> Result<(), rumqttc::v5::ClientError> {
+        let dest = Ipv4Header::from_slice(&packet.get_bytes())
+            .ok()
+            .and_then(|(ipv4_header, _)| {
+                let tuple: (u8, u8, u8, u8) = ipv4_header.destination.into();
```

Comment:
Isn't it easier to just get the IP as a `u32`? Or are there endianess issues? I know network stuff are big-endian but both x86 and ARM uses small-endian...

Reply:
ipv4_header.destination is a `[u8;4]` and I was looking for a function to turn it into an Ipv4Addr, the [IntoAddress trait](https://docs.rs/tun/latest/tun/trait.IntoAddress.html) doesn't have a impl for `[u8;4]` but there's one for tuple so I just tried to convert it, except, now I found there's a [From](https://doc.rust-lang.org/nightly/core/net/struct.Ipv4Addr.html#impl-From%3C%5Bu8;+4%5D%3E-for-Ipv4Addr) impl between them, so everything is good